File: webhook-deliveries-stuck-failed/python/retry_stale_failed_deliveries.py
"""Find Saleor webhook EventDeliveries stuck FAILED past Celery's retry limit
and retry only the ones that look worth retrying.

send_webhook_request (saleor/plugins/webhook/tasks.py) retries an async
delivery with retry_backoff=10 and retry_kwargs={"max_retries": 5}, roughly
10 * 2^n seconds of delay across 5 attempts. Once the fifth retry also
fails, Celery never reschedules the task again and the EventDelivery is
persisted as FAILED for good. Nothing in Saleor resurrects it later.

Under DRY_RUN=true (the default) this only reports what it would do. When
DRY_RUN=false it calls eventDeliveryRetry once per delivery classified
RETRY, and re-polls its status to confirm it left FAILED. Deliveries whose
recent attempts all look like a dead endpoint (5xx or no response code)
are only ever reported, never retried. Run on a schedule. Safe to run
again and again.

Guide: https://www.allanninal.dev/saleor/webhook-deliveries-stuck-failed/
"""
import os
import time
import logging
import datetime
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("retry_stale_failed_deliveries")

API_URL = os.environ["SALEOR_API_URL"]
TOKEN = os.environ["SALEOR_AUTH_TOKEN"]
STALE_AFTER_MINUTES = float(os.environ.get("STALE_AFTER_MINUTES", "60"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

DEFAULT_MAX_RETRIES = 5
DEFAULT_STALE_AFTER_MS = 3600000  # 1 hour, Saleor's Celery max_retries=5

FAILED_DELIVERIES_QUERY = """
query {
  webhooks(first: 50) {
    edges {
      node {
        id
        name
        isActive
        targetUrl
        eventDeliveries(first: 100, filter: { status: FAILED },
                         sortBy: { field: CREATED_AT, direction: DESC }) {
          edges {
            node {
              id
              eventType
              createdAt
              status
              attempts(first: 5, sortBy: { field: CREATED_AT, direction: DESC }) {
                edges { node { id createdAt status duration responseStatusCode taskId } }
              }
            }
          }
        }
      }
    }
  }
}"""

DELIVERY_STATUS_QUERY = """
query($id: ID!) {
  eventDelivery(id: $id) { id status }
}"""

RETRY_MUTATION = """
mutation($id: ID!) {
  eventDeliveryRetry(id: $id) {
    delivery { id status }
    errors { field code message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        API_URL,
        json={"query": query, "variables": variables or {}},
        headers={"Authorization": f"Bearer {TOKEN}", "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def _parse_iso_ms(iso):
    return datetime.datetime.fromisoformat(iso.replace("Z", "+00:00")).timestamp() * 1000


def _looks_dead(attempt):
    code = attempt.get("responseStatusCode")
    return code is None or code >= 500


def decide_stale_failed_retries(deliveries, now_iso, opts=None):
    """Pure decision function: no I/O, fully unit-testable.

    deliveries: list of {id, status, createdAt, attempts: [{createdAt, responseStatusCode}]}
    now_iso: ISO 8601 timestamp string
    opts: optional {maxRetries, staleAfterMs}

    Returns a list of {id, action, reason} where action is one of
    'RETRY', 'FLAG_DEAD_ENDPOINT', 'SKIP'.
    """
    opts = opts or {}
    max_retries = opts.get("maxRetries", DEFAULT_MAX_RETRIES)
    stale_after_ms = opts.get("staleAfterMs", DEFAULT_STALE_AFTER_MS)
    now_ms = _parse_iso_ms(now_iso)

    decisions = []
    for delivery in deliveries:
        if delivery.get("status") != "FAILED":
            decisions.append({"id": delivery["id"], "action": "SKIP", "reason": "not-failed"})
            continue

        attempts = delivery.get("attempts") or []
        attempt_count = len(attempts)
        last_attempt_at = delivery.get("createdAt")
        if attempts:
            last_attempt_at = max(a["createdAt"] for a in attempts)
        age_ms = now_ms - _parse_iso_ms(last_attempt_at)

        if attempt_count < max_retries and age_ms < stale_after_ms:
            # Saleor's own Celery retries may still be in flight, don't race them.
            decisions.append({"id": delivery["id"], "action": "SKIP", "reason": "still-within-retry-window"})
            continue

        if attempt_count >= max_retries and age_ms >= stale_after_ms:
            recent = attempts[:max_retries]
            all_dead = all(_looks_dead(a) for a in recent) if recent else False
            if all_dead:
                decisions.append({"id": delivery["id"], "action": "FLAG_DEAD_ENDPOINT",
                                   "reason": "endpoint-repeatedly-unreachable"})
            else:
                decisions.append({"id": delivery["id"], "action": "RETRY",
                                   "reason": "stale-failed-past-retry-limit-transient-error"})
            continue

        decisions.append({"id": delivery["id"], "action": "SKIP",
                           "reason": "recently-exhausted-wait-for-staleness-window"})

    return decisions


def failed_deliveries_by_webhook():
    data = gql(FAILED_DELIVERIES_QUERY)["webhooks"]
    for edge in data["edges"]:
        webhook = edge["node"]
        deliveries = [d["node"] for d in webhook["eventDeliveries"]["edges"]]
        yield webhook, deliveries


def retry_delivery(delivery_id):
    result = gql(RETRY_MUTATION, {"id": delivery_id})["eventDeliveryRetry"]
    if result["errors"]:
        raise RuntimeError(result["errors"])
    return result["delivery"]["status"]


def confirm_left_failed(delivery_id):
    data = gql(DELIVERY_STATUS_QUERY, {"id": delivery_id})["eventDelivery"]
    return data["status"] if data else None


def run():
    now_iso = datetime.datetime.now(datetime.timezone.utc).isoformat()
    opts = {"staleAfterMs": STALE_AFTER_MINUTES * 60 * 1000}

    retried = 0
    flagged = 0
    for webhook, deliveries in failed_deliveries_by_webhook():
        decisions = decide_stale_failed_retries(deliveries, now_iso, opts)
        by_id = {d["id"]: d for d in deliveries}

        for decision in decisions:
            if decision["action"] == "SKIP":
                continue

            delivery = by_id[decision["id"]]

            if decision["action"] == "FLAG_DEAD_ENDPOINT":
                codes = [a.get("responseStatusCode") for a in (delivery.get("attempts") or [])]
                log.warning(
                    "DEAD ENDPOINT webhook=%s (%s) delivery=%s eventType=%s recent_codes=%s",
                    webhook["name"], webhook["targetUrl"], delivery["id"], delivery["eventType"], codes,
                )
                flagged += 1
                continue

            if decision["action"] == "RETRY":
                if DRY_RUN:
                    log.info("[DRY RUN] would retry delivery %s (%s, %s)",
                              delivery["id"], delivery["eventType"], webhook["name"])
                else:
                    log.info("Retrying delivery %s (%s, %s)",
                              delivery["id"], delivery["eventType"], webhook["name"])
                    retry_delivery(delivery["id"])
                    time.sleep(2)
                    status = confirm_left_failed(delivery["id"])
                    log.info("Delivery %s status after retry: %s", delivery["id"], status)
                retried += 1

    log.info(
        "Done. %d delivery(ies) %s, %d dead endpoint(s) flagged for review.",
        retried, "to retry" if DRY_RUN else "retried", flagged,
    )


if __name__ == "__main__":
    run()

File: webhook-deliveries-stuck-failed/node/retry-stale-failed-deliveries.js
/**
 * Find Saleor webhook EventDeliveries stuck FAILED past Celery's retry limit
 * and retry only the ones that look worth retrying.
 *
 * send_webhook_request (saleor/plugins/webhook/tasks.py) retries an async
 * delivery with retry_backoff=10 and retry_kwargs={"max_retries": 5}, roughly
 * 10 * 2^n seconds of delay across 5 attempts. Once the fifth retry also
 * fails, Celery never reschedules the task again and the EventDelivery is
 * persisted as FAILED for good. Nothing in Saleor resurrects it later.
 *
 * Under DRY_RUN=true (the default) this only reports what it would do. When
 * DRY_RUN=false it calls eventDeliveryRetry once per delivery classified
 * RETRY, and re-polls its status to confirm it left FAILED. Deliveries whose
 * recent attempts all look like a dead endpoint (5xx or no response code)
 * are only ever reported, never retried. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/saleor/webhook-deliveries-stuck-failed/
 */
import { pathToFileURL } from "node:url";

const API_URL = process.env.SALEOR_API_URL || "https://store.saleor.cloud/graphql/";
const TOKEN = process.env.SALEOR_AUTH_TOKEN || "dummy-token";
const STALE_AFTER_MINUTES = Number(process.env.STALE_AFTER_MINUTES || 60);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const DEFAULT_MAX_RETRIES = 5;
const DEFAULT_STALE_AFTER_MS = 3600000; // 1 hour, Saleor's Celery max_retries=5

/**
 * Pure decision function: no I/O, fully unit-testable.
 *
 * @param {Array<{id: string, status: string, createdAt: string, attempts: Array<{createdAt: string, responseStatusCode?: number}>}>} deliveries
 * @param {string} nowIso
 * @param {{maxRetries?: number, staleAfterMs?: number}} [opts]
 * @returns {Array<{id: string, action: 'RETRY'|'FLAG_DEAD_ENDPOINT'|'SKIP', reason: string}>}
 */
export function decideStaleFailedRetries(deliveries, nowIso, opts = {}) {
  const maxRetries = opts.maxRetries ?? DEFAULT_MAX_RETRIES;
  const staleAfterMs = opts.staleAfterMs ?? DEFAULT_STALE_AFTER_MS;
  const nowMs = Date.parse(nowIso);

  return deliveries.map((delivery) => {
    if (delivery.status !== "FAILED") {
      return { id: delivery.id, action: "SKIP", reason: "not-failed" };
    }

    const attempts = delivery.attempts || [];
    const attemptCount = attempts.length;
    const lastAttemptAt = attempts.length
      ? attempts.reduce((max, a) => (a.createdAt > max ? a.createdAt : max), attempts[0].createdAt)
      : delivery.createdAt;
    const ageMs = nowMs - Date.parse(lastAttemptAt);

    if (attemptCount < maxRetries && ageMs < staleAfterMs) {
      // Saleor's own Celery retries may still be in flight, don't race them.
      return { id: delivery.id, action: "SKIP", reason: "still-within-retry-window" };
    }

    if (attemptCount >= maxRetries && ageMs >= staleAfterMs) {
      const recent = attempts.slice(0, maxRetries);
      const allDead = recent.length > 0 && recent.every(looksDead);
      return allDead
        ? { id: delivery.id, action: "FLAG_DEAD_ENDPOINT", reason: "endpoint-repeatedly-unreachable" }
        : { id: delivery.id, action: "RETRY", reason: "stale-failed-past-retry-limit-transient-error" };
    }

    return { id: delivery.id, action: "SKIP", reason: "recently-exhausted-wait-for-staleness-window" };
  });
}

function looksDead(attempt) {
  const code = attempt.responseStatusCode;
  return code === null || code === undefined || code >= 500;
}

async function gql(query, variables = {}) {
  const res = await fetch(API_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${TOKEN}`, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Saleor ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const FAILED_DELIVERIES_QUERY = `
query {
  webhooks(first: 50) {
    edges {
      node {
        id
        name
        isActive
        targetUrl
        eventDeliveries(first: 100, filter: { status: FAILED },
                         sortBy: { field: CREATED_AT, direction: DESC }) {
          edges {
            node {
              id
              eventType
              createdAt
              status
              attempts(first: 5, sortBy: { field: CREATED_AT, direction: DESC }) {
                edges { node { id createdAt status duration responseStatusCode taskId } }
              }
            }
          }
        }
      }
    }
  }
}`;

const DELIVERY_STATUS_QUERY = `
query($id: ID!) {
  eventDelivery(id: $id) { id status }
}`;

const RETRY_MUTATION = `
mutation($id: ID!) {
  eventDeliveryRetry(id: $id) {
    delivery { id status }
    errors { field code message }
  }
}`;

async function* failedDeliveriesByWebhook() {
  const data = (await gql(FAILED_DELIVERIES_QUERY)).webhooks;
  for (const edge of data.edges) {
    const webhook = edge.node;
    const deliveries = webhook.eventDeliveries.edges.map((d) => d.node);
    yield { webhook, deliveries };
  }
}

async function retryDelivery(deliveryId) {
  const result = (await gql(RETRY_MUTATION, { id: deliveryId })).eventDeliveryRetry;
  if (result.errors.length) throw new Error(JSON.stringify(result.errors));
  return result.delivery.status;
}

async function confirmLeftFailed(deliveryId) {
  const data = (await gql(DELIVERY_STATUS_QUERY, { id: deliveryId })).eventDelivery;
  return data ? data.status : null;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function run() {
  const nowIso = new Date().toISOString();
  const opts = { staleAfterMs: STALE_AFTER_MINUTES * 60 * 1000 };

  let retried = 0;
  let flagged = 0;

  for await (const { webhook, deliveries } of failedDeliveriesByWebhook()) {
    const decisions = decideStaleFailedRetries(deliveries, nowIso, opts);
    const byId = new Map(deliveries.map((d) => [d.id, d]));

    for (const decision of decisions) {
      if (decision.action === "SKIP") continue;
      const delivery = byId.get(decision.id);

      if (decision.action === "FLAG_DEAD_ENDPOINT") {
        const codes = (delivery.attempts || []).map((a) => a.responseStatusCode);
        console.warn(
          `DEAD ENDPOINT webhook=${webhook.name} (${webhook.targetUrl}) delivery=${delivery.id} eventType=${delivery.eventType} recent_codes=${JSON.stringify(codes)}`
        );
        flagged++;
        continue;
      }

      if (decision.action === "RETRY") {
        console.log(
          DRY_RUN
            ? `[DRY RUN] would retry delivery ${delivery.id} (${delivery.eventType}, ${webhook.name})`
            : `Retrying delivery ${delivery.id} (${delivery.eventType}, ${webhook.name})`
        );
        if (!DRY_RUN) {
          await retryDelivery(delivery.id);
          await sleep(2000);
          const status = await confirmLeftFailed(delivery.id);
          console.log(`Delivery ${delivery.id} status after retry: ${status}`);
        }
        retried++;
      }
    }
  }

  console.log(
    `Done. ${retried} delivery(ies) ${DRY_RUN ? "to retry" : "retried"}, ${flagged} dead endpoint(s) flagged for review.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
